libraries/AP_Torqeedo/AP_Torqeedo.cpp

```diff
@@ -0,0 +1,310 @@
+void AP_Torqeedo::init()
+{
+    // only init once
+    if (_initialised) {
+        return;
+    }
+
+    // create background thread to process serial input and output
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Torqeedo::loop, void), "torqeedo", 4096, AP_HAL::Scheduler::PRIORITY_RCOUT, 1)) {
+        return;
+    }
+}
+
+// initialise serial port and gpio pins (run from background thread)
+void AP_Torqeedo::init_internals()
+{
+    // find serial driver and initialise
+    const AP_SerialManager &serial_manager = AP::serialmanager();
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Torqeedo, 0);
+    if (_uart == nullptr) {
+        return;
+    }
+    _uart->begin(TORQEEDO_SERIAL_BAUD);
+    _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+    _uart->set_unbuffered_writes(true);
+
+    // initialise onoff pin and set for 0.5 sec to turn on motor
+    if (_pin_onoff > -1) {
+        hal.gpio->pinMode(_pin_onoff, HAL_GPIO_OUTPUT);
+        hal.gpio->write(_pin_onoff, 1);
+
+        // delay 0.5 sec and then unset pin
+        hal.scheduler->delay(500);
+        hal.gpio->write(_pin_onoff, 0);
+    }
+
+    // initialise RS485 DE pin (when high, allows send to motor)
+    if (_pin_de > -1) {
+        hal.gpio->pinMode(_pin_de, HAL_GPIO_OUTPUT);
+        hal.gpio->write(_pin_de, 0);
+    }
+
+    _initialised = true;
+}
+
+// consume incoming messages from motor, reply with latest motor speed
+// runs in background thread
+void AP_Torqeedo::loop()
+{
+    // initialisation
+    init_internals();
+    if (!_initialised) {
+        return;
+    }
+
+    while (true) {
+        // 1ms loop delay
+        hal.scheduler->delay_microseconds(1000);
+
+        // check if transmit pin should be unset
+        check_for_send_end();
+
+        // parse incoming characters
+        uint32_t nbytes = MIN(_uart->available(), 1024U);
+        bool motor_speed_request_received = false;
+        while (nbytes-- > 0) {
+            int16_t b = _uart->read();
+            if (b >= 0 ) {
+                if (parse_byte((uint8_t)b)) {
+                    // request received to send updated motor speed
+                    motor_speed_request_received = true;
+                }
+                parse_bytes_received++;
+            }
+        }
+
+        // send motor speed
+        if (motor_speed_request_received) {
+            send_motor_speed_cmd();
+        }
+    }
+}
+
+// process a single byte received on serial port
+// return true if a this driver should send a set-motor-speed message
+bool AP_Torqeedo::parse_byte(uint8_t b)
+{
+    bool motor_speed_requested = false;
+
+    switch (_parse_state) {
+    case ParseState::WAITING_FOR_HEADER:
+        if (b == TORQEEDO_PACKET_HEADER) {
+            _parse_state = ParseState::WAITING_FOR_FOOTER;
+        }
+        received_buff_len = 0;
+        break;
+    case ParseState::WAITING_FOR_FOOTER:
+        if (b == TORQEEDO_PACKET_FOOTER) {
```

Comment:
I don't think this parser is sufficient - at the very least the CRC could match the packet footer, in which case you'll try to parse a packet which is shorter than expected (and you'll probably lose the next packet, too.

I'd suggest something like the "move-header-in-buffer" style parser on new serial parsers, as used in the RichenPower generator.

Reply:
I see what you're saying and you're right that if the packet has "AD" in it anywhere we will lose the packet.  Sadly the packets are all different lengths and I have not deciphered what they are all for yet nor how to distinguish them from each other as they arrive (i.e. I don't know where the message id field is yet).

I think I will leave this one open for now but I accept that it is a real problem.

Reply:
tridge suggests this might work by adding some timing constraints